src/databricks/labs/ucx/hive_metastore/pipelines_migrate.py

```diff
@@ -0,0 +1,160 @@
+import logging
+from dataclasses import dataclass
+from functools import partial
+from typing import BinaryIO
+from collections.abc import Generator
+
+from databricks.labs.blueprint.parallel import Threads
+from databricks.labs.lsql.backends import SqlBackend
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors import NotFound, DatabricksError
+from databricks.labs.blueprint.installation import Installation
+
+from databricks.labs.ucx.assessment.pipelines import PipelinesCrawler, PipelineInfo
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class PipelineRule:
+    workspace_name: str
+    src_pipeline_id: str
+    target_catalog_name: str | None = None
+    target_schema_name: str | None = None
+    target_pipeline_name: str | None = None
+
+    @classmethod
+    def from_src_dst(
+        cls,
+        workspace_name: str,
+        src_pipeline_id: str,
+        target_catalog_name: str | None = None,
+        target_schema_name: str | None = None,
+        target_pipeline_name: str | None = None,
+    ) -> "PipelineRule":
+        return cls(
+            workspace_name=workspace_name,
+            src_pipeline_id=src_pipeline_id,
+            target_catalog_name=target_catalog_name,
+            target_schema_name=target_schema_name,
+            target_pipeline_name=target_pipeline_name,
+        )
+
+    @classmethod
+    def initial(cls, workspace_name: str, catalog_name: str, pipeline: PipelineInfo) -> "PipelineRule":
+        return cls(
+            workspace_name=workspace_name,
+            target_catalog_name=catalog_name,
+            src_pipeline_id=pipeline.pipeline_id,
+            target_pipeline_name=pipeline.pipeline_name,
+            target_schema_name=None,
+        )
+
+
+@dataclass
+class PipelineToMigrate:
+    src: PipelineInfo
+    rule: PipelineRule
+
+    def __hash__(self):
```

Comment:
Where are these dunder methods used? I expect you can use the `dataclasses.field` instead: https://docs.python.org/3/library/dataclasses.html#dataclasses.field.

``` python
@dataclass
class PipelineToMigrate:
    src: PipelineInfo
    rule: PipelineRule = field(compare=False)
```